File: Cpp/Array/448_Find_All_Missing_Numbers.cpp
#include <vector>
#include <unordered_set>

using namespace std;

// Brute Force Solution - O(n log n)
class Solution_Brute {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        
        vector<int> missing;
        int expected = 1;
        
        for (int num : nums) {
            // Gap - fill missing numbers
            while (num > expected) {
                missing.push_back(expected);
                expected++;
            }
            // Matched expected
            if (num == expected) {
                expected++;
            }
        }
        
        // Add remaining missing numbers
        while (expected <= nums.size()) {
            missing.push_back(expected);
            expected++;
        }
        
        return missing;
    }
};

// Optimal Solution - Hash Set O(n)
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        unordered_set<int> numSet(nums.begin(), nums.end());
        vector<int> result;
        
        for (int i = 1; i <= nums.size(); i++) {
            if (numSet.find(i) == numSet.end()) {
                result.push_back(i);
            }
        }
        
        return result;
    }
};

// Alternative Optimal Solution - In-place marking
class Solution_InPlace {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        // Mark numbers as negative to indicate they exist
        for (int num : nums) {
            int index = abs(num) - 1;
            if (nums[index] > 0) {
                nums[index] = -nums[index];
            }
        }
        
        vector<int> result;
        
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) {
                result.push_back(i + 1);
            }
        }
        
        return result;
    }
};

// Local Testing
int main() {
    Solution sol;
    Solution_Brute sol_brute;
    Solution_InPlace sol_inplace;
    
    vector<int> array1 = {4, 3, 2, 7, 8, 2, 3, 1};
    
    vector<int> result = sol.findDisappearedNumbers(array1);
    
    return 0;
}

File: Java/Array/448_Find_All_Missing_Numbers.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class SolutionBrute {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        int[] sorted = nums.clone();
        java.util.Arrays.sort(sorted);
        List<Integer> result = new ArrayList<>();
        int expected = 1;

        for (int num : sorted) {
            while (num > expected) result.add(expected++);
            if (num == expected) expected++;
        }
        while (expected <= nums.length) result.add(expected++);
        return result;
    }
}

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        Set<Integer> present = new HashSet<>();
        for (int num : nums) present.add(num);

        List<Integer> result = new ArrayList<>();
        for (int value = 1; value <= nums.length; value++) {
            if (!present.contains(value)) result.add(value);
        }
        return result;
    }
}

class SolutionInPlace {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        for (int num : nums) {
            int index = Math.abs(num) - 1;
            if (nums[index] > 0) nums[index] = -nums[index];
        }

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) result.add(i + 1);
        }
        return result;
    }
}
